src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonNode.cs

```diff
@@ -0,0 +1,248 @@
+
+        /// <summary>
+        ///   Options to control the behavior.
+        /// </summary>
+        public JsonNodeOptions? Options
+        {
+            get
+            {
+                if (!_options.HasValue && Parent != null)
+                {
+                    _options = Parent.Options;
+                }
+
+                return _options;
+            }
+
+            private set
+            {
+                _options = value;
+            }
+        }
+
+        internal JsonNode(JsonNodeOptions? options = null)
+        {
+            Options = options;
+        }
+
+        /// <summary>
+        ///   Casts to the derived <see cref="JsonArray"/> type.
+        /// </summary>
+        /// <returns>
+        ///   A <see cref="JsonArray"/>.
+        /// </returns>
+        /// <exception cref="InvalidOperationException">
+        ///   The node is not a <see cref="JsonArray"/>.
+        /// </exception>
+        public JsonArray AsArray()
+        {
+            if (this is JsonArray jArray)
+            {
+                return jArray;
+            }
+
+            throw new InvalidOperationException(SR.Format(SR.NodeWrongType, nameof(JsonArray)));
+        }
+
+        /// <summary>
+        ///   Casts to the derived <see cref="JsonObject"/> type.
+        /// </summary>
+        /// <returns>
+        ///   A <see cref="JsonObject"/>.
+        /// </returns>
+        /// <exception cref="InvalidOperationException">
+        ///   The node is not a <see cref="JsonObject"/>.
+        /// </exception>
+        public JsonObject AsObject()
+        {
+            if (this is JsonObject jObject)
+            {
+                return jObject;
+            }
+
+            throw new InvalidOperationException(SR.Format(SR.NodeWrongType, nameof(JsonObject)));
+        }
+
+        /// <summary>
+        ///   Casts to the derived <see cref="JsonValue"/> type.
+        /// </summary>
+        /// <returns>
+        ///   A <see cref="JsonValue"/>.
+        /// </returns>
+        /// <exception cref="InvalidOperationException">
+        ///   The node is not a <see cref="JsonValue"/>.
+        /// </exception>
+        public JsonValue AsValue()
+        {
+            if (this is JsonValue jValue)
+            {
+                return jValue;
+            }
+
+            throw new InvalidOperationException(SR.Format(SR.NodeWrongType, nameof(JsonValue)));
+        }
+
+        /// <summary>
+        ///   Gets the parent <see cref="JsonNode"/>.
+        ///   If there is no parent, <see langword="null"/> is returned.
+        ///   A parent can either be a <see cref="JsonObject"/> or a <see cref="JsonArray"/>.
+        /// </summary>
+        public JsonNode? Parent
+        {
+            get
+            {
+                return _parent;
+            }
+            internal set
+            {
+                _parent = value;
+            }
+        }
+
+        /// <summary>
+        ///   Gets the JSON path.
```

Comment:
I believe I saw a similar formatter defined in either `WriteStack` or `ReadStack`. Would it make sense to consolidate into a shared helper method?

Reply:
The code is different since they operate on different inputs (JsonNode vs WriteStack\ReadStack). Perhaps a strategy or visitor pattern can be applied however although that may be overkill.